### ISSUE TITLE
RangeError: Maximum call stack size exceeded with deepEqualExlcudingEvery

### ISSUE THREAD

**Issue by sanex3339** [mesaugat/chai-exclude#14]
I got `RangeError: Maximum call stack size exceeded` with `deepEqualExlcudingEvery`
Deep equal working fine, without `Maximum call stack size exceeded`

Call:
```assert.deepEqualExcludingEvery(structure, expectedStructure, ['start', 'end']);```

Expected Object:
```
  {
    "declarations": [
      {
        "id": {
          "metadata": {
            "ignoredNode": false
          }
          "name": "abc"
          "parentNode": [Circular]
          "type": "Identifier"
        }
        "init": {
          "metadata": {
            "ignoredNode": false
          }
          "parentNode": [Circular]
          "raw": "'cde'"
          "type": "Literal"
          "value": "cde"
          "x-verbatim-property": {
            "content": "'cde'"
            "precedence": 20
          }
        }
        "metadata": {
          "ignoredNode": false
        }
        "parentNode": {
          "declarations": [Circular]
          "kind": "var"
          "metadata": {
            "ignoredNode": false
          }
          "parentNode": {
            "body": [
              [Circular]
            ]
            "metadata": {
              "ignoredNode": false
            }
            "parentNode": [Circular]
            "sourceType": "script"
            "type": "Program"
          }
          "type": "VariableDeclaration"
        }
        "type": "VariableDeclarator"
      }
    ]
    "kind": "var"
    "metadata": {
      "ignoredNode": false
    }
    "parentNode": {
      "body": [
        {
          "declarations": [
            {
              "id": {
                "metadata": {
                  "ignoredNode": false
                }
                "name": "abc"
                "parentNode": [Circular]
                "type": "Identifier"
              }
              "init": {
                "metadata": {
                  "ignoredNode": false
                }
                "parentNode": [Circular]
                "raw": "'cde'"
                "type": "Literal"
                "value": "cde"
                "x-verbatim-property": {
                  "content": "'cde'"
                  "precedence": 20
                }
              }
              "metadata": {
                "ignoredNode": false
              }
              "parentNode": [Circular]
              "type": "VariableDeclarator"
            }
          ]
          "kind": "var"
          "metadata": {
            "ignoredNode": false
          }
          "parentNode": [Circular]
          "type": "VariableDeclaration"
        }
      ]
      "metadata": {
        "ignoredNode": false
      }
      "parentNode": [Circular]
      "sourceType": "script"
      "type": "Program"
    }
    "type": "VariableDeclaration"
  }
```

Actual Object:
```
  {
    "declarations": [
      {
        "end": 32
        "id": {
          "end": 24
          "metadata": {
            "ignoredNode": false
          }
          "name": "abc"
          "parentNode": [Circular]
          "start": 21
          "type": "Identifier"
        }
        "init": {
          "end": 32
          "metadata": {
            "ignoredNode": false
          }
          "parentNode": [Circular]
          "raw": "'cde'"
          "start": 27
          "type": "Literal"
          "value": "cde"
          "x-verbatim-property": {
            "content": "'cde'"
            "precedence": 20
          }
        }
        "metadata": {
          "ignoredNode": false
        }
        "parentNode": {
          "declarations": [Circular]
          "end": 33
          "kind": "var"
          "metadata": {
            "ignoredNode": false
          }
          "parentNode": {
            "body": [
              [Circular]
            ]
            "end": 46
            "metadata": {
              "ignoredNode": false
            }
            "parentNode": [Circular]
            "sourceType": "script"
            "start": 0
            "type": "Program"
          }
          "start": 17
          "type": "VariableDeclaration"
        }
        "start": 21
        "type": "VariableDeclarator"
      }
    ]
    "kind": "var"
    "metadata": {
      "ignoredNode": false
    }
    "parentNode": {
      "body": [
        {
          "declarations": [
            {
              "end": 32
              "id": {
                "end": 24
                "metadata": {
                  "ignoredNode": false
                }
                "name": "abc"
                "parentNode": [Circular]
                "start": 21
                "type": "Identifier"
              }
              "init": {
                "end": 32
                "metadata": {
                  "ignoredNode": false
                }
                "parentNode": [Circular]
                "raw": "'cde'"
                "start": 27
                "type": "Literal"
                "value": "cde"
                "x-verbatim-property": {
                  "content": "'cde'"
                  "precedence": 20
                }
              }
              "metadata": {
                "ignoredNode": false
              }
              "parentNode": [Circular]
              "start": 21
              "type": "VariableDeclarator"
            }
          ]
          "end": 33
          "kind": "var"
          "metadata": {
            "ignoredNode": false
          }
          "parentNode": [Circular]
          "start": 17
          "type": "VariableDeclaration"
        }
      ]
      "end": 46
      "metadata": {
        "ignoredNode": false
      }
      "parentNode": [Circular]
      "sourceType": "script"
      "start": 0
      "type": "Program"
    }
    "type": "VariableDeclaration"
  }
```

**Comment by mesaugat** [comment]
I think this is because of the Circular reference. I'll look into it and get back to you when possible.

**Comment by mesaugat** [comment]
@sanex3339 I think this is an AST tree of some sort. Can you please post a valid JavaScript object that you are actually using.

**Comment by mesaugat** [comment]
```js
const actual = {
    "declarations": [
      {
        "end": 32,
        "id": {
          "end": 24,
          "metadata": {
            "ignoredNode": false
          },
          "name": "abc",
          "parentNode": "[Circular]",
          "start": 21,
          "type": "Identifier"
        },
        "init": {
          "end": 32,
          "metadata": {
            "ignoredNode": false
          },
          "parentNode": "[Circular]",
          "raw": "cde",
          "start": 27,
          "type": "Literal",
          "value": "cde",
          "x-verbatim-property": {
            "content": "cde",
            "precedence": 20
          }
        },
        "metadata": {
          "ignoredNode": false
        },
        "parentNode": {
          "declarations": "[Circular]",
          "end": 33,
          "kind": "var",
          "metadata": {
            "ignoredNode": false
          },
          "parentNode": {
            "body": [
              "[Circular]",
            ],
            "end": 46,
            "metadata": {
              "ignoredNode": false
            },
            "parentNode": "[Circular]",
            "sourceType": "script",
            "start": 0,
            "type": "Program"
          },
          "start": 17,
          "type": "VariableDeclaration"
        },
        "start": 21,
        "type": "VariableDeclarator"
      }
    ],
    "kind": "var",
    "metadata": {
      "ignoredNode": false
    },
    "parentNode": {
      "body": [
        {
          "declarations": [
            {
              "end": 32,
              "id": {
                "end": 24,
                "metadata": {
                  "ignoredNode": false
                },
                "name": "abc",
                "parentNode": "[Circular]",
                "start": 21,
                "type": "Identifier"
              },
              "init": {
                "end": 32,
                "metadata": {
                  "ignoredNode": false
                },
                "parentNode": "[Circular]",
                "raw": "cde",
                "start": 27,
                "type": "Literal",
                "value": "cde",
                "x-verbatim-property": {
                  "content": "cde",
                  "precedence": 20
                }
              },
              "metadata": {
                "ignoredNode": false
              },
              "parentNode": "[Circular]",
              "start": 21,
              "type": "VariableDeclarator"
            }
          ],
          "end": 33,
          "kind": "var",
          "metadata": {
            "ignoredNode": false
          },
          "parentNode": "[Circular]",
          "start": 17,
          "type": "VariableDeclaration"
        }
      ],
      "end": 46,
      "metadata": {
        "ignoredNode": false
      },
      "parentNode": "[Circular]",
      "sourceType": "script",
      "start": 0,
      "type": "Program",
    },
    "type": "VariableDeclaration"
  }

  const expected = {
    "declarations": [
      {
        "id": {
          "metadata": {
            "ignoredNode": false
          },
          "name": "abc",
          "parentNode": "[Circular]",
          "type": "Identifier"
        },
        "init": {
          "metadata": {
            "ignoredNode": false
          },
          "parentNode": "[Circular]",
          "raw": "cde",
          "type": "Literal",
          "value": "cde",
          "x-verbatim-property": {
            "content": "cde",
            "precedence": 20
          }
        },
        "metadata": {
          "ignoredNode": false
        },
        "parentNode": {
          "declarations": "[Circular]",
          "kind": "var",
          "metadata": {
            "ignoredNode": false
          },
          "parentNode": {
            "body": [
              "[Circular]"
            ],
            "metadata": {
              "ignoredNode": false
            },
            "parentNode": "[Circular]",
            "sourceType": "script",
            "type": "Program"
          },
          "type": "VariableDeclaration"
        },
        "type": "VariableDeclarator"
      }
    ],
    "kind": "var",
    "metadata": {
      "ignoredNode": false
    },
    "parentNode": {
      "body": [
        {
          "declarations": [
            {
              "id": {
                "metadata": {
                  "ignoredNode": false
                },
                "name": "abc",
                "parentNode": "[Circular]",
                "type": "Identifier"
              },
              "init": {
                "metadata": {
                  "ignoredNode": false
                },
                "parentNode": "[Circular]",
                "raw": "cde",
                "type": "Literal",
                "value": "cde",
                "x-verbatim-property": {
                  "content": "cde",
                  "precedence": 20
                }
              },
              "metadata": {
                "ignoredNode": false
              },
              "parentNode": "[Circular]",
              "type": "VariableDeclarator"
            }
          ],
          "kind": "var",
          "metadata": {
            "ignoredNode": false
          },
          "parentNode": "[Circular]",
          "type": "VariableDeclaration"
        }
      ],
      "metadata": {
        "ignoredNode": false
      },
      "parentNode": "[Circular]",
      "sourceType": "script",
      "type": "Program"
    },
    "type": "VariableDeclaration"
  }

assert.deepEqualExcludingEvery(actual, expected, ['start', 'end'])
```

This is working fine.

**Comment by mesaugat** [comment]
Hello @sanex3339, is this still an issue?

**Comment by sanex3339** [comment]
Hi, i think, issue still exists, but i don't using chai exclude anymore

**Comment by mesaugat** [comment]
@sanex3339 Can you provide me the code snippet that was not working so that I can figure out a solution?

**Comment by mesaugat** [comment]
Closing this because I am not sure what the problem is. If someone is able to replicate this issue please comment down below with the failing scenario.

**Comment by atlassubbed** [comment]
I'm running into the same issue, this library doesn't handle circular dependencies. Might want to add some test cases for circular references.

@mesaugat All you did with OP's example was change the circular references to strings. Do you think that solves the problem? Lol.

@sanex3339 What are you using instead?

**Comment by mesaugat** [comment]
@atlassubbed Yes, it doesn't work with circular references. 

It should work when you change circular references to string.

**Comment by atlassubbed** [comment]
You might want to update the `README.md` to clarify that this is "unsafe" (doesn't handle circular refs), otherwise people might assume it behaves like Chai's deep equal, which gracefully handles circular references. Since you're plugging into chai, you might want to handle circular refs in this library. 

I don't think people will be able to convert their circular refs to strings if they aren't able to filter their objects deeply in the first place. Anyway, like OP, I just wrote my own function to do the field filtering, so my issue has been solved.

Edit: Here's an example I found of what I mean by "unsafe" https://github.com/moll/json-stringify-safe
